File: web/src/services/api/series_asset_lock.ts
// novel-workflow v2: series-asset-lock (漫剧一致性) API client
// 后端路由: /api/v1/novel/series-asset-lock/*

import { apiGet, apiPost, apiPut } from './request';

export interface SeriesAssetLock {
  id: string;
  userId: string;
  projectId: string;
  characterIdsJson: string;
  sceneIdsJson: string;
  propIdsJson: string;
  globalStylePrompt: string;
  isLocked: boolean;
  lockedAt: string;
  unlockedAt: string;
  createdAt: string;
  updatedAt: string;
}

// 解析后的主资产包（前端易用视图）
export interface SeriesAssetLockParsed {
  id: string;
  projectId: string;
  characterIds: string[];
  sceneIds: string[];
  propIds: string[];
  globalStylePrompt: string;
  isLocked: boolean;
  lockedAt: string;
}

// GET /api/v1/novel/series-asset-lock?projectId=...
export async function getSeriesAssetLock(projectId: string): Promise<SeriesAssetLock | null> {
  return apiGet<SeriesAssetLock | null>('/api/v1/novel/series-asset-lock', { projectId });
}

// PUT /api/v1/novel/series-asset-lock?projectId=...
export async function updateSeriesAssetLock(req: {
  projectId: string;
  characterIds: string[];
  sceneIds: string[];
  propIds: string[];
  globalStylePrompt: string;
}): Promise<SeriesAssetLock> {
  return apiPut<SeriesAssetLock>(
    `/api/v1/novel/series-asset-lock?projectId=${encodeURIComponent(req.projectId)}`,
    {
      characterIds: req.characterIds,
      sceneIds: req.sceneIds,
      propIds: req.propIds,
      globalStylePrompt: req.globalStylePrompt,
    },
  );
}

// POST /api/v1/novel/series-asset-lock/lock?projectId=...
export async function lockSeriesAssetLock(projectId: string): Promise<SeriesAssetLock> {
  return apiPost<SeriesAssetLock>(
    `/api/v1/novel/series-asset-lock/lock?projectId=${encodeURIComponent(projectId)}`,
  );
}

// POST /api/v1/novel/series-asset-lock/unlock?projectId=...
export async function unlockSeriesAssetLock(projectId: string): Promise<SeriesAssetLock> {
  return apiPost<SeriesAssetLock>(
    `/api/v1/novel/series-asset-lock/unlock?projectId=${encodeURIComponent(projectId)}`,
  );
}

// 工具：把后端 JSON 解析为 string[] 视图
export function parseSeriesAssetLock(lock: SeriesAssetLock | null): SeriesAssetLockParsed | null {
  if (!lock) return null;
  return {
    id: lock.id,
    projectId: lock.projectId,
    characterIds: lock.characterIdsJson ? JSON.parse(lock.characterIdsJson) : [],
    sceneIds: lock.sceneIdsJson ? JSON.parse(lock.sceneIdsJson) : [],
    propIds: lock.propIdsJson ? JSON.parse(lock.propIdsJson) : [],
    globalStylePrompt: lock.globalStylePrompt,
    isLocked: lock.isLocked,
    lockedAt: lock.lockedAt,
  };
}

File: model/series_asset_lock.go
package model

// SeriesAssetLock novel-workflow v2 - series-asset-lock (漫剧一致性) 的项目级主资产包。
//
// 概念：在项目级别锁定一组"主资产包"：1-N 个角色、1-N 个场景、1-N 个道具、
// 一段全局色调 prompt。视频生成强制从主资产包取参考图；
// 跨分镜共享风格 reference；保证漫剧 / 短剧的世界观一致性。
//
// 锁定后：
//   - 分镜级资产选择器只显示主资产包内的资产
//   - 所有分镜视频生成的 prompt 自动追加全局色调描述
//   - 同场景分镜共享同一张场景图
type SeriesAssetLock struct {
	ID        string `json:"id" gorm:"primaryKey"`
	UserID    string `json:"userId" gorm:"index"`
	ProjectID string `json:"projectId" gorm:"index"`

	// 主资产包内容
	CharacterIDsJSON string `json:"characterIdsJson" gorm:"type:text"`
	SceneIDsJSON     string `json:"sceneIdsJson" gorm:"type:text"`
	PropIDsJSON      string `json:"propIdsJson" gorm:"type:text"`

	// 全局色调 prompt
	GlobalStylePrompt string `json:"globalStylePrompt" gorm:"type:text"`

	// 是否锁定
	IsLocked bool `json:"isLocked"`

	// 锁定时间
	LockedAt   string `json:"lockedAt"`
	UnlockedAt string `json:"unlockedAt"`

	CreatedAt string `json:"createdAt" gorm:"index"`
	UpdatedAt string `json:"updatedAt" gorm:"index"`
}
